feat: UI Dialog and DialogTrigger Components (#953)
deephaven/deephaven-plugins: plugins/ui/docs/components/dialog_trigger.md

# Dialog Trigger

Dialog trigger serves as a wrapper around a dialog and its associated trigger, linking the dialog's open state with the trigger's press state. Additionally, it allows you to customize the type and positioning of the dialog.

## Example

```python
from deephaven import ui

my_dialog_trigger_example = ui.dialog_trigger(
    ui.action_button(
        "Disk Status",
    ),
    ui.dialog(
        ui.heading("C://"),
        ui.content("50% disk space remaining"),
    ),
    type="popover",
)
```

## Content

The `dialog_trigger` accepts exactly two children: the element which triggers opening of the `dialog` and the `dialog` itself. The trigger must be the first child passed into the `dialog_trigger` and should be an element that supports press events.

> If your `dialog` has buttons within it that should close the Dialog when pressed, you must use controlled mode in order to propagate the close function to the dialog's children. Dialogs that do not contain such interactive elements can simply provide the `dialog` component as is to the `dialog_trigger` as its second child.

Reviewer: ```suggestion
If your `dialog` has buttons within it that should close the dialog when pressed, you must use controlled mode to propagate the close function to the dialog's children. Dialogs that do not contain such interactive elements can provide the `dialog` component as is to the `dialog_trigger` as its second child.
```

The example below demonstrates how to pass the close function to the dialog's buttons.

```python
from deephaven import ui

@ui.component
def close_example():
    is_open, set_open, set_closed = ui.use_flag()
    return ui.dialog_trigger(
        ui.action_button("Check connectivity", on_press=set_open),
        ui.dialog(
            ui.heading("Internet Speed Test"),
            ui.content("Start speed test?"),
            ui.button_group(
                ui.button("Cancel", variant="secondary", on_press=set_closed),
                ui.button("Confirm", variant="accent", on_press=set_closed),
            ),
        ),
        is_open=is_open,
    )

my_close_example = close_example()
```

## Dialog types

By providing a `type` prop, you can specify the type of `dialog` that is rendered by your `dialog_trigger`. Note that pressing the `Esc` key will close the `dialog` regardless of its type.

### Modal

> Modal dialogs create an underlay that blocks access to the underlying user interface until the dialog is closed. Sizing options can be found on the `dialog` page. Focus is trapped inside the Modal as per the accessibility guidelines laid out by W3C.

Reviewer: ```suggestion
Modal dialogs create an underlay that blocks access to the underlying user interface until the dialog is closed. Sizing options can be found on the `dialog` page. Focus is trapped inside the modal as per the accessibility guidelines laid out by W3C.
```

        type="fullscreen",
    )

my_fullscreen_example = fullscreen_example()
```

### Fullscreen takeover

Fullscreen takeover dialogs are similar to the fullscreen variant except that the `dialog` covers the entire screen.

```python
from deephaven import ui

@ui.component
def fullscreen_takeover_example():
    is_open, set_open, set_closed = ui.use_flag()
    return ui.dialog_trigger(
        ui.action_button("Trigger Fullscreen", on_press=set_open),
        ui.dialog(
            ui.heading("Fullscreen"),
            ui.content(
                ui.form(
                    ui.text_field(label="Name"),
                    ui.text_field(label="Email address"),
                    ui.checkbox("Make profile private"),
                )
            ),
            ui.button_group(
                ui.button("Cancel", variant="secondary", on_press=set_closed),
                ui.button(
                    "Confirm", variant="accent", on_press=set_closed, auto_focus=True
                ),
            ),
        ),
        is_open=is_open,
        type="fullscreenTakeover",
    )

my_fullscreen_takeover_example = fullscreen_takeover_example()
```

### Dismissable

> If your modal fialog doesn't require the user to make a confirmation, you can set `is_dismissable` on the `dialog_trigger`. This adds a close button that the user can press to dismiss the `dialog`.

Reviewer 2: ```suggestion
If your modal dialog doesn't require the user to make a confirmation, you can set `is_dismissable` on the `dialog_trigger`. This adds a close button that the user can press to dismiss the `dialog`.
```

```python
from deephaven import ui

my_dialog_example2 = ui.dialog_trigger(
    ui.action_button("Status"),
    ui.dialog(ui.heading("Status"), ui.content("Printer Status: Connected")),
    is_dismissable=True,
)
```

## Dialog placement

Popover dialogs support a variety of placement options since they do not take over the user interface like modal or tray dialogs.

### Placement

The popover's placement can be adjusted using the `placement` prop. The full list of placements is:

- `bottom`
- `bottom left`
- `bottom right`
- `bottom start`
- `bottom end`
- `top`
- `top left`
- `top right`
- `top start`
- `top end`
- `left`
- `left top`
- `left bottom`
- `start`
- `start top`
- `start bottom`
- `right`
- `right top`
- `right bottom`
- `end`
- `end top`
- `end bottom`

```python
from deephaven import ui

my_placement = ui.dialog_trigger(
    ui.action_button("Trigger"),
    ui.dialog(
        ui.heading("The Heading"),
        ui.content(
            "This is a popover placed to the right of its\
          trigger and offset so the arrow is at the top of the dialog."
        ),
    ),
    type="popover",
    placement="right top",
)
```

### Offset and cross offset

> The popover dialog's offset can be adjusted using the `offset` and `cross_offset` props. The `offset` prop controls the spacing applied along the main axis between the element and its anchor element whereas the `cross_offset` prop handles the spacing applied along the cross axis.

Reviewer: ```suggestion
The popover dialog's offset can be adjusted using the `offset` and `cross_offset` props. The `offset` prop controls the spacing applied along the main axis between the element and its anchor element, whereas the `cross_offset` prop handles the spacing applied along the cross axis.
```

```python
from deephaven import ui

my_cross_offset = ui.dialog_trigger(
    ui.action_button("Trigger"),
    ui.dialog(
        ui.heading("Cross offset"),
        ui.content("Cross offset by an additional 100px."),
    ),
    type="popover",
    cross_offset=100,
)
```

### Flipping

By default, `dialog_trigger` attempts to flip popovers on the main axis in situations where the original placement would cause it to render out of view. This can be overridden by setting `should_flip` to `False`.

```python
from deephaven import ui

my_should_flip_example = ui.dialog_trigger(
    ui.action_button("Trigger"),
    ui.dialog(
        ui.heading("The Heading"),
        ui.content(
            "This is a popover that won't flip if it can't fully render below the button."
        ),
    ),
    type="popover",
    placement="bottom",
    should_flip=False,
)
```

### Container padding

You can control the minimum padding required between the popover dialog and the surrounding container via the `container_padding` prop. This affects the positioning breakpoints that determine when the dialog will attempt to flip.

> The example below will flip the dialog from above the trigger button to below the trigger button if the Dialog cannot render fully while maintaining 50px of padding between itself and the top of the browser.

Reviewer: ```suggestion
The example below will flip the dialog from above the trigger button to below the trigger button if the dialog cannot render fully while maintaining 50px of padding between itself and the top of the browser.
```

Reviewer: or `dialog`?

```python
from deephaven import ui

my_should_flip_example = ui.dialog_trigger(
    ui.action_button("Trigger"),
    ui.dialog(
        ui.heading("The Heading"),
        ui.content("This is a popover."),
    ),
    type="popover",
    placement="top",
    container_padding=50,
)
```

## Events

> Dialog trigger accepts an `on_open_change` handler which is triggered whenever the Dialog is opened or closed.

Reviewer: ```suggestion
Dialog triggers accept an `on_open_change` handler that is triggered whenever the dialog is opened or closed. The example below updates a separate element with the dialog's current open state.

```

> The example below uses `on_open_change` to update a separate element with the current open state of the Dialog.

Reviewer: ```suggestion
```

> ```python

Reviewer: pausing here
